modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java

```diff
@@ -610,18 +623,211 @@ static Tuple<Long, Long> numberOfMultiparts(final long totalSize, final long par
+
+            if (uploadIndex < 0) {
+                // already aborted by someone else
+                listener.onResponse(OptionalLong.empty());
+                return;
+            }
+
+            final var isComplete = new AtomicBoolean();
+            final Runnable doCleanup = () -> {
+                if (isComplete.compareAndSet(false, true)) {
+                    try {
+                        abortMultipartUploadIfExists(uploadId);
+                    } catch (Exception e) {
+                        // cleanup is a best-effort thing, we can't do anything better than log and fall through here
+                        logger.error("unexpected error cleaning up upload [" + uploadId + "] of [" + blobKey + "]", e);
+                        assert false : e;
+                    }
+                }
+            };
+
+            try (
+                var listeners = new RefCountingListener(
+                    ActionListener.runAfter(
+                        listener.delegateFailure(
+                            (delegate1, ignored) -> getRegister(
+                                rawKey,
+                                delegate1.delegateFailure((delegate2, currentValue) -> ActionListener.completeWith(delegate2, () -> {
+                                    if (currentValue.isPresent() && currentValue.getAsLong() == expected) {
+                                        final var completeMultipartUploadRequest = new CompleteMultipartUploadRequest(
+                                            bucket,
+                                            blobKey,
+                                            uploadId,
+                                            List.of(partETag)
+                                        );
+                                        completeMultipartUploadRequest.setRequestMetricCollector(blobStore.multiPartUploadMetricCollector);
+                                        SocketAccess.doPrivilegedVoid(() -> client.completeMultipartUpload(completeMultipartUploadRequest));
+                                        isComplete.set(true);
+                                    }
+                                    return currentValue;
+                                }))
+                            )
+                        ),
+                        doCleanup
+                    )
+                )
+            ) {
+                if (currentUploads.size() > 1) {
+                    // This is a small optimization to improve the liveness properties of this algorithm.
+                    //
+                    // When there are multiple competing updates, we order them by upload id and the first one tries to cancel the competing
+                    // updates in order to make progress. To avoid liveness issues when the winner fails, the rest wait based on their
+                    // upload_id-based position and try to make progress.
+
+                    var delayListener = listeners.acquire();
```

Comment:
I think there's a risk of releasing all the `listeners` refs before the multipart upload tasks are dispatched into the `SNAPSHOT` thread pool if?

Reply:
We `.acquire()` a listener for each upload while constructing the `ActionRunnable` to be dispatched, and that all happens on this thread before completing `delayListener`.

Reply:
You're right, thanks for the clarification 👍